contracts/chain-adapters/Solana_Adapter.sol

```diff
@@ -0,0 +1,193 @@
+        // Solana adapter requires CCTP TokenMessenger and MessageTransmitter contracts to be set.
+        if (address(_cctpTokenMessenger) == address(0)) {
+            revert InvalidCctpTokenMessenger(address(_cctpTokenMessenger));
+        }
+        if (address(_cctpMessageTransmitter) == address(0)) {
+            revert InvalidCctpMessageTransmitter(address(_cctpMessageTransmitter));
+        }
+
+        cctpMessageTransmitter = _cctpMessageTransmitter;
+
+        SOLANA_SPOKE_POOL_BYTES32 = solanaSpokePool;
+        SOLANA_SPOKE_POOL_ADDRESS = _trimSolanaAddress(solanaSpokePool);
+
+        SOLANA_USDC_BYTES32 = solanaUsdc;
+        SOLANA_USDC_ADDRESS = _trimSolanaAddress(solanaUsdc);
+
+        SOLANA_SPOKE_POOL_USDC_VAULT = solanaSpokePoolUsdcVault;
+    }
+
+    /**
+     * @notice Send cross-chain message to target on Solana.
+     * @dev Only allows sending messages to the Solana spoke pool.
+     * @param target Program on Solana (translated as EVM address) that will receive message.
+     * @param message Data to send to target.
+     */
+    function relayMessage(address target, bytes calldata message) external payable override {
+        if (target != SOLANA_SPOKE_POOL_ADDRESS) {
+            revert InvalidRelayMessageTarget(target);
+        }
+
+        bytes4 selector = bytes4(message[:4]);
+        if (selector == SpokePoolInterface.setEnableRoute.selector) {
+            cctpMessageTransmitter.sendMessage(
+                CircleDomainIds.Solana,
+                SOLANA_SPOKE_POOL_BYTES32,
+                _translateSetEnableRoute(message)
+            );
+        } else {
+            cctpMessageTransmitter.sendMessage(CircleDomainIds.Solana, SOLANA_SPOKE_POOL_BYTES32, message);
+        }
+
+        // TODO: consider if we need also to emit the translated message.
+        emit MessageRelayed(target, message);
+    }
+
+    /**
+     * @notice Bridge tokens to Solana.
+     * @dev Only allows bridging USDC to Solana spoke pool.
+     * @param l1Token L1 token to deposit.
+     * @param l2Token L2 token to receive.
+     * @param amount Amount of L1 tokens to deposit and L2 tokens to receive.
+     * @param to Bridge recipient.
+     */
+    function relayTokens(
+        address l1Token,
+        address l2Token,
+        uint256 amount,
+        address to
+    ) external payable override {
+        if (l1Token != address(usdcToken)) {
+            revert InvalidL1Token(l1Token);
+        }
+        if (l2Token != SOLANA_USDC_ADDRESS) {
+            revert InvalidL2Token(l2Token);
+        }
+        if (amount > type(uint64).max) {
+            revert InvalidAmount(amount);
+        }
+        if (to != SOLANA_SPOKE_POOL_ADDRESS) {
+            revert InvalidTokenRecipient(to);
+        }
+
+        _transferUsdc(SOLANA_SPOKE_POOL_USDC_VAULT, amount);
+
+        // TODO: consider if we need also to emit the translated addresses.
+        emit TokensRelayed(l1Token, l2Token, amount, to);
```

Comment:
I think it could be useful to emit the translated addresses at very litttle additional marginal cost